File: sample/cs/14_raymarching_light_cycle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Any = System.Object;
using int64 = System.Int64;
using float64 = System.Double;
using str = System.String;
using Pytra.CsModule;
using math = Pytra.CsModule.math;

public static class Program
{
    // 14: Sample that outputs a moving-light scene in a simple raymarching style as a GIF.
    
    public static List<byte> palette()
    {
        List<byte> p = new System.Collections.Generic.List<byte>();
        long i = 0;
        for (i = 0; i < 256; i += 1) {
            long r = System.Math.Min(255, Pytra.CsModule.py_runtime.py_int(20 + i * 0.9));
            long g = System.Math.Min(255, Pytra.CsModule.py_runtime.py_int(10 + i * 0.7));
            long b = System.Math.Min(255, 30 + i);
            Pytra.CsModule.py_runtime.py_append(p, r);
            Pytra.CsModule.py_runtime.py_append(p, g);
            Pytra.CsModule.py_runtime.py_append(p, b);
        }
        return Pytra.CsModule.py_runtime.py_bytes(p);
    }
    
    public static long scene(double x, double y, double light_x, double light_y)
    {
        double x1 = x + 0.45;
        double y1 = y + 0.2;
        double x2 = x - 0.35;
        double y2 = y - 0.15;
        var r1 = math.sqrt(x1 * x1 + y1 * y1);
        var r2 = math.sqrt(x2 * x2 + y2 * y2);
        var blob = math.exp(-7.0 * r1 * r1) + math.exp(-8.0 * r2 * r2);
        
        double lx = x - light_x;
        double ly = y - light_y;
        var l = math.sqrt(lx * lx + ly * ly);
        double lit = System.Convert.ToDouble(1.0) / System.Convert.ToDouble((1.0 + 3.5 * l * l));
        
        long v = Pytra.CsModule.py_runtime.py_int(255.0 * blob * lit * 5.0);
        return System.Math.Min(255, System.Math.Max(0, v));
    }
    
    public static void run_14_raymarching_light_cycle()
    {
        long w = 320;
        long h = 240;
        long frames_n = 84;
        string out_path = "sample/out/14_raymarching_light_cycle.gif";
        
        double start = Pytra.CsModule.time.perf_counter();
        System.Collections.Generic.List<List<byte>> frames = new System.Collections.Generic.List<List<byte>>();
        
        long t = 0;
        for (t = 0; t < frames_n; t += 1) {
            List<byte> frame = Pytra.CsModule.py_runtime.py_bytearray(w * h);
            double a = (System.Convert.ToDouble(t) / System.Convert.ToDouble(frames_n)) * math.pi * 2.0;
            double light_x = 0.75 * math.cos(a);
            double light_y = 0.55 * math.sin(a * 1.2);
            
            long y = 0;
            for (y = 0; y < h; y += 1) {
                long row_base = y * w;
                double py = (System.Convert.ToDouble(y) / System.Convert.ToDouble((h - 1))) * 2.0 - 1.0;
                long x = 0;
                for (x = 0; x < w; x += 1) {
                    double px = (System.Convert.ToDouble(x) / System.Convert.ToDouble((w - 1))) * 2.0 - 1.0;
                    Pytra.CsModule.py_runtime.py_set(frame, row_base + x, scene(px, py, light_x, light_y));
                }
            }
            frames.Add(Pytra.CsModule.py_runtime.py_bytes(frame));
        }
        Pytra.CsModule.gif_helper.save_gif(out_path, w, h, frames, palette(), 3, 0);
        double elapsed = Pytra.CsModule.time.perf_counter() - start;
        System.Console.WriteLine(string.Join(" ", new object[] { "output:", out_path }));
        System.Console.WriteLine(string.Join(" ", new object[] { "frames:", frames_n }));
        System.Console.WriteLine(string.Join(" ", new object[] { "elapsed_sec:", elapsed }));
    }
    
    public static void Main(string[] args)
    {
            run_14_raymarching_light_cycle();
    }
}

File: sample/ruby/14_raymarching_light_cycle.rb
require_relative "built_in/py_runtime"
require_relative "std/math"
require_relative "std/time"
require_relative "utils/gif"


# 14: Sample that outputs a moving-light scene in a simple raymarching style as a GIF.

def palette()
  p = __pytra_bytearray()
  i = 0
  while i < 256
    r = __pytra_min(255, __pytra_int((20 + i * 0.9)))
    g = __pytra_min(255, __pytra_int((10 + i * 0.7)))
    b = __pytra_min(255, 30 + i)
    p.concat([r, g, b])
    i += 1
  end
  return __pytra_bytes(p)
end

def scene(x, y, light_x, light_y)
  x1 = x + 0.45
  y1 = y + 0.2
  x2 = x - 0.35
  y2 = y - 0.15
  r1 = sqrt((x1 * x1 + y1 * y1))
  r2 = sqrt((x2 * x2 + y2 * y2))
  blob = exp((((-7.0) * r1) * r1)) + exp((((-8.0) * r2) * r2))
  lx = x - light_x
  ly = y - light_y
  l = sqrt((lx * lx + ly * ly))
  lit = __pytra_div(1.0, (1.0 + (3.5 * l * l)))
  v = __pytra_int(((255.0 * blob * lit) * 5.0))
  return __pytra_min(255, __pytra_max(0, v))
end

def run_14_raymarching_light_cycle()
  w = 320
  h = 240
  frames_n = 84
  out_path = "sample/out/14_raymarching_light_cycle.gif"
  start = perf_counter()
  frames = []
  t = 0
  while t < frames_n
    frame = __pytra_bytearray(w * h)
    a = ((__pytra_div(t, frames_n) * pi) * 2.0)
    light_x = 0.75 * cos(a)
    light_y = 0.55 * sin(a * 1.2)
    y = 0
    while y < h
      row_base = y * w
      py = ((__pytra_div(y, (h - 1)) * 2.0) - 1.0)
      x = 0
      while x < w
        px = ((__pytra_div(x, (w - 1)) * 2.0) - 1.0)
        __pytra_set_index(frame, row_base + x, scene(px, py, light_x, light_y))
        x += 1
      end
      y += 1
    end
    frames.append(__pytra_bytes(frame))
    t += 1
  end
  save_gif(out_path, w, h, frames, palette(), 3, 0)
  elapsed = perf_counter() - start
  __pytra_print("output:", out_path)
  __pytra_print("frames:", frames_n)
  __pytra_print("elapsed_sec:", elapsed)
end

if __FILE__ == $PROGRAM_NAME
  run_14_raymarching_light_cycle()
end
